resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/service/StreamingResourceServiceProxy.java

```diff
@@ -85,83 +55,86 @@ public StreamingResourceServiceProxy(final PersistenceLayer persistence, final R
         this.persistence = persistence;
         this.delegate = delegate;
         this.objectMapper = objectMapper;
-        this.resourceBuilder = Collections::emptyList;
     }
 
     /**
-     * Construct a StreamingResourceServiceProxy, passing each member as an argument
+     * Uses a resource to get any {@link LeafResource}s associated with the it.
      *
-     * @param persistence     a {@link PersistenceLayer} for persisting resources in, as if it were a cache
-     * @param delegate        a 'real' {@link ResourceService} to delegate requests to when not found in the persistence layer
-     * @param objectMapper    a {@link ObjectMapper} used for serialisation when writing each {@link Resource} to the {@link java.io.OutputStream}
-     * @param resourceBuilder a {@link Supplier} of resources as built by a {@link uk.gov.gchq.palisade.service.ResourcePrepopulationFactory}, but with a connection detail attached
+     * @param resource  the resource to be queried
+     * @return          a {@link FunctionalIterator} of {@link LeafResource}s associated with the resource
      */
-    public StreamingResourceServiceProxy(final PersistenceLayer persistence, final ResourceService delegate, final ObjectMapper objectMapper, final Supplier<List<Entry<Resource, LeafResource>>> resourceBuilder) {
-        this(persistence, delegate, objectMapper);
-        this.resourceBuilder = resourceBuilder;
-    }
-
     @Transactional
-    public void getResourcesByResource(final Resource resource, final OutputStream outputStream) {
+    public FunctionalIterator<LeafResource> getResourcesByResource(final Resource resource) {
         // Try first from persistence
-        LOGGER.info("Trying from persistence store");
-        Stream<LeafResource> resourceStream = persistence.getResourcesById(resource.getId())
-                // Otherwise call out to resource service
-                .orElseGet(() -> {
-                    LOGGER.info("Persistence empty, delegating to resource-service");
-                    // Persist returned resources as the stream is consumed
-                    return persistence.withPersistenceById(resource.getId(), delegate.getResourcesById(resource.getId()));
-                });
-        // Consume the stream, write to the output stream
-        serialiseAndWriteStreamToOutput(resourceStream, outputStream);
+        LOGGER.info(STORE);
+        FunctionalIterator<LeafResource> persistenceIterator = FunctionalIterator.fromIterator(persistence.getResourcesById(resource.getId()));
+        if (persistenceIterator.hasNext()) {
+            return FunctionalIterator.fromIterator(persistenceIterator);
+        } else {
+            LOGGER.info(EMPTY);
+            return persistence.withPersistenceById(resource.getId(),
+                    FunctionalIterator.fromIterator(delegate.getResourcesById(resource.getId())));
+        }
```

Comment:
This is the case for the optional returned, we want to differentiate between:
* cache hit -> empty dir -> empty iterator
* cache miss -> call resource-service

In this implementation, the resource-service is needlessly called for empty dirs

Reply:
Implemented the above work flow.